src/libAtomVM/posix_nifs.c

```diff
@@ -363,6 +383,92 @@ static term nif_atomvm_posix_write(Context *ctx, int argc, term argv[])
 
     return result;
 }
+
+static term nif_atomvm_posix_select(Context *ctx, term argv[], enum ErlNifSelectFlags mode)
+{
+    term process_pid_term = argv[1];
+    VALIDATE_VALUE(process_pid_term, term_is_pid);
+    int32_t process_pid = term_to_local_process_id(process_pid_term);
+    term select_ref_term = argv[2];
+    if (select_ref_term != UNDEFINED_ATOM) {
+        VALIDATE_VALUE(select_ref_term, term_is_reference);
+    }
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), argv[0], ctx->global->posix_fd_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct PosixFd *fd_obj = (struct PosixFd *) fd_obj_ptr;
+
+    if (UNLIKELY(enif_select(erl_nif_env_from_context(ctx), fd_obj->fd, mode, fd_obj, &process_pid, select_ref_term) < 0)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    fd_obj->select = 1;
+
+    return OK_ATOM;
+}
+
+static term nif_atomvm_posix_select_read(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    return nif_atomvm_posix_select(ctx, argv, ERL_NIF_SELECT_READ);
+}
+
+static term nif_atomvm_posix_select_write(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    return nif_atomvm_posix_select(ctx, argv, ERL_NIF_SELECT_WRITE);
+}
+
+static term nif_atomvm_posix_select_stop(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), argv[0], ctx->global->posix_fd_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct PosixFd *fd_obj = (struct PosixFd *) fd_obj_ptr;
+    if (UNLIKELY(enif_select(erl_nif_env_from_context(ctx), fd_obj->fd, ERL_NIF_SELECT_STOP, fd_obj, NULL, term_nil()) < 0)) {
+        RAISE_ERROR(BADARG_ATOM);
```

Comment:
I realize that it might be beyond the scope of this PR, but it looks like the OTP `enif_select` encodes reasons in the negative return value when there is an error.  So I wonder if it makes some sense to try to pick those apart a raise a different reason.

Reply:
OTP's documentation on `enif_select` mentions `ERL_NIF_SELECT_INVALID_EVENT` and `ERL_NIF_SELECT_FAILED`. We define both, but our implementation doesn't return `ERL_NIF_SELECT_FAILED`.
```suggestion
        RAISE_ERROR(BADARG_ATOM);
```
```suggestion
        RAISE_ERROR(BADARG_ATOM);
```
```suggestion
        RAISE_ERROR(BADARG_ATOM);
```
```suggestion
        RAISE_ERROR(BADARG_ATOM);
```

Eventually, I had a look at OTP's enif select implementation and it is very different regarding the returned value.

The implementation here will return `ERL_NIF_SELECT_INVALID_EVENT` if the event cannot be found when calling select with stop. OTP's just accept it [silently](https://github.com/erlang/otp/blob/1d9faf5a52cb947886b2d16a8ea21cbd911a6e41/erts/emulator/sys/common/erl_check_io.c#L1106).

OTP checks, under some circumstances, that the event is positive, and returns `ERL_NIF_SELECT_INVALID_EVENT` if it is negative. We don't, but since select is platform dependent it's probably better.

OTP [asserts if mode is 0](https://github.com/erlang/otp/blob/1d9faf5a52cb947886b2d16a8ea21cbd911a6e41/erts/emulator/sys/common/erl_check_io.c#L1116), we return `ERL_NIF_SELECT_BADARG` which is an AtomVM addition.

OTP returns `ERL_NIF_SELECT_FAILED` if it cannot add the event to the poll set. We could modify the signature of `sys_register_select_event` to get an error and likewise report it with `ERL_NIF_SELECT_FAILED`. With kqueue, we do add to the set synchronously and we currently abort if it fails. But with epoll, we do it asynchronously.

Should we align behavior with OTP's? In particular, should `enif_select` be verbose about issues and tolerate bad arguments (e.g. without asserting?). Also should we distinguish in nif_atomvm_posix_select* between all these cases? I explicitely chose `badarg` here because it seemed to be the most common exception when an argument is wrong.

Reply:
WFM